### PR TITLE
Support group router for moe models

### PR DESCRIPTION
**Pull request by RunningLeon**
## Motivation

Support group router for moe models

### pipeline
```python
from lmdeploy import pipeline, GenerationConfig, PytorchEngineConfig

if __name__ == '__main__':
    backend_config = PytorchEngineConfig(hf_overrides=dict(router_n_groups=4))
    model_path = 'Qwen/Qwen3-30B-A3B'
    pipe = pipeline(model_path, backend_config=backend_config)

    resps = pipe(['Hi.'])
    for res in resps:
        print(res)
```

### api server

```shell
lmdeploy serve api_server \
Qwen/Qwen3-30B-A3B \
--backend pytorch \
--hf-overrides '{"router_n_groups": 4}'
```
## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Review by lvhan028 — APPROVED** [review]

File: lmdeploy/pytorch/backends/moe.py [hunk 1]
```diff
@@ -20,7 +20,7 @@ class SoftmaxTopKBuilder(ABC):
 
     @staticmethod
     @abstractmethod
-    def build(top_k: int, dim: int = -1):
+    def build(top_k: int, dim: int = -1, n_groups: int = -1):
```
Comment (grimoire) [inline lmdeploy/pytorch/backends/moe.py:30]: Dlinfer backend failed.

Comment (grimoire) [reply lmdeploy/pytorch/backends/moe.py:30]: Maybe it is better to add new method to SoftmaxTopKImpl, so we do not to enable the feature on init.
```
class SoftmaxTopKImpl:
    def __init__():
         self.n_groups=-1

    def set_n_groups(self, n_groups):
         self.n_groups = n_groups

```

So we don't have to update other backend

File: lmdeploy/pytorch/backends/default/moe.py [hunk 2]
```diff
@@ -7,21 +7,35 @@
...
         self.top_k = top_k
         self.dim = dim
+        self.n_groups = n_groups
 
     def forward(self, x: torch.Tensor):
         """forward."""
         routing_weights = torch.softmax(x, dim=self.dim, dtype=torch.float32)
-        topk_weights, topk_ids = torch.topk(routing_weights, self.top_k, dim=self.dim)
+        if self.n_groups > 0:
+            assert self.top_k % self.n_groups == 0, f'{self.top_k} cannot be divided by {self.n_groups}'
```
Comment (grimoire) [inline lmdeploy/pytorch/backends/default/moe.py]: Why don't we assert it in init?

File: lmdeploy/pytorch/backends/default/moe.py [hunk 3]
```diff
@@ -7,21 +7,35 @@
...
+            assert routing_weights.shape[
```
Comment (grimoire) [inline lmdeploy/pytorch/backends/default/moe.py:20]: `shape[-1]` should be `shape[self.dim]` ?

File: lmdeploy/pytorch/backends/default/moe.py [hunk 4]
```diff
@@ -7,21 +7,35 @@
...
+                -1] % self.n_groups == 0, f'{routing_weights.shape[-1]} cannot be divided by {self.n_groups}'
+            per_group_top_k = self.top_k // self.n_groups
+            group_size = routing_weights.shape[-1] // self.n_groups
+            group_offsets = (torch.arange(self.n_groups, device=routing_weights.device) * group_size).view(
```
Comment (grimoire) [inline lmdeploy/pytorch/backends/default/moe.py]: Since  n_groups is fixed, we can create a function with lru_cache

```python
@functools.lru_cache
def get_group_offsets(n_groups, group_size, device):
     return ....
```

So we can reuse the same group_offsets
